library/rsa.c

```diff
@@ -1429,15 +1442,27 @@ int mbedtls_rsa_rsassa_pkcs1_v15_verify( mbedtls_rsa_context *ctx,
     if( asn1_len != hashlen )
         return( MBEDTLS_ERR_RSA_VERIFY_FAILED );
 
-    if( memcmp( p, hash, hashlen ) != 0 )
-        return( MBEDTLS_ERR_RSA_VERIFY_FAILED );
+    ret = MBEDTLS_ERR_RSA_VERIFY_FAILED;
+    ret = memcmp( p, hash, hashlen );
 
-    p += hashlen;
+    if( ret == 0 )
+    {
+        if( ret != 0)
+            return( MBEDTLS_ERR_RSA_VERIFY_FAILED );
 
-    if( p != end )
-        return( MBEDTLS_ERR_RSA_VERIFY_FAILED );
+        p += hashlen;
+        vp = p;
 
-    return( 0 );
+        if( vp == end )
+        {
+            if( vp != end )
```

Comment:
You're testing that `vp != end`, directly after asserting the opposite, that `vp == end`. That looks like a logical contradiction to me.